src/Features/Core/Portable/FileHeaders/AbstractFileHeaderHelper.cs

```diff
@@ -0,0 +1,152 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+
+namespace Microsoft.CodeAnalysis.FileHeaders
+{
+    internal abstract class AbstractFileHeaderHelper
+    {
+        protected AbstractFileHeaderHelper(ISyntaxKinds syntaxKinds)
+        {
+            SingleLineCommentTriviaKind = syntaxKinds.SingleLineCommentTrivia;
+            MultiLineCommentTriviaKind = syntaxKinds.MultiLineCommentTrivia;
+            WhitespaceTriviaKind = syntaxKinds.WhitespaceTrivia;
+            EndOfLineTriviaKind = syntaxKinds.EndOfLineTrivia;
+        }
+
+        public abstract string CommentPrefix { get; }
+
+        private int SingleLineCommentTriviaKind { get; }
+        private int MultiLineCommentTriviaKind { get; }
+        private int WhitespaceTriviaKind { get; }
+        private int EndOfLineTriviaKind { get; }
+
+        protected abstract string GetTextContextOfComment(SyntaxTrivia commentTrivia);
+
+        public FileHeader ParseFileHeader(SyntaxNode root)
+        {
+            var firstToken = root.GetFirstToken(includeZeroWidth: true);
+            var firstNonWhitespaceTrivia = IndexOfFirstNonWhitespaceTrivia(firstToken.LeadingTrivia, true);
+
+            if (firstNonWhitespaceTrivia == -1)
+            {
+                return FileHeader.MissingFileHeader(0);
+            }
+
+            var sb = StringBuilderPool.Allocate();
```

Comment:
`using var _ = PooledStringBuilder.Allocate(out var sb)`?  then you don't have to remember the frees.